platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt

```diff
@@ -142,149 +146,120 @@ sealed class PropertyTrace {
+        override fun describe(builder: StructuredMessage.Builder) = builder.apply {
+            text("system property ")
+            reference(name)
+            text(" set at ")
+        }
     }
 
-    abstract override fun toString(): String
-
     abstract override fun equals(other: Any?): Boolean
 
     abstract override fun hashCode(): Int
 
+    abstract override fun toString(): String
+
     /**
      * The shared logic for implementing `toString()` in subclasses.
      */
     protected
     fun asString(): String =
-        StringBuilder().apply {
+        StructuredMessage.Builder().apply {
             sequence.forEach {
-                appendStringOf(it)
+                it.describe(this)
             }
-        }.toString()
+        }.build().asString('`')
 
     /**
      * The user code where the problem occurred. User code should generally be some coarse-grained entity such as a plugin or script.
      */
     open val containingUserCode: String
-        get() = StringBuilder().apply {
-            appendStringOf(this@PropertyTrace)
-        }.toString()
-
-    private
-    fun StringBuilder.appendStringOf(trace: PropertyTrace) {
-        when (trace) {
-            is Gradle -> {
-                append("Gradle runtime")
-            }
-
-            is Property -> {
-                append(trace.kind)
-                append(" ")
-                quoted(trace.name)
-                append(" of ")
-            }
-
-            is SystemProperty -> {
-                append("system property ")
-                quoted(trace.name)
-                append(" set at ")
-            }
-
-            is Bean -> {
-                quoted(trace.type.name)
-                append(" bean found in ")
-            }
-
-            is Task -> {
-                append("task ")
-                quoted(trace.path)
-                append(" of type ")
-                quoted(trace.type.name)
-            }
-            is BuildLogic -> {
-                append(trace.source.displayName)
-                trace.lineNumber?.let {
-                    append(": line ")
-                    append(it)
-                }
-            }
-
-            is BuildLogicClass -> {
-                append("class ")
-                quoted(trace.name)
-            }
-
-            is Unknown -> {
-                append("unknown location")
-            }
+        get() = containingUserCodeMessage.asString('`')
 
-            is Project -> {
-                append("project ")
-                quoted(trace.path)
-                append(" in ")
-            }
-        }
-    }
+    open val containingUserCodeMessage: StructuredMessage
+        get() = describe(StructuredMessage.Builder()).build()
 
-    private
-    fun StringBuilder.quoted(s: String) {
-        append('`')
-        append(s)
-        append('`')
-    }
+    abstract fun describe(builder: StructuredMessage.Builder): StructuredMessage.Builder
```

Comment:
💅  More idiomatic would be to make `StructuredMessage.Builder` a receiver for this function. Then you don't need `builder.apply` in any overrides. Then the return type can also be `Unit`

Reply:
IMHO we use extension functions a bit too much, leading to unexpected direction of dependencies. Between `StructuredMessage.Builder` and `PropertyTrace`, I think it makes sense for a method in the latter to take the former as an argument, but not the other way around (`PropertyTrace` may know about `StructuredMessage`, but not the other way around). (gets off soapbox)

Reply:
I'm not suggesting to make `StructuredMessage` know about `PropertyTrace`.

Here is what I meant:
```
        abstract fun StructuredMessage.Builder.describe()

        override fun StructuredMessage.Builder.describe() {
            text("$kind ")
            reference(name)
            text(" of ")
        }
```

It's making `PropertyTrace` describe itself in a context of the `StructuredMessage`. Similar pattern for how `CoroutineScope` is supposed to be used in Kotlin.

Reply:
Yup, I get that, but I see extension methods as emulating methods that could be hosted by the receiver - that is what it looks like from call sites. I understand we could (as a convention) make it mean it whatever we wanted though. I am fine with we changing it later to be a receiver, but as an argument it seems to work just as well, and it seems to reflect the responsibilities of each type better. I chose to use `with()` blocks to set it as the default receiver _within_ each implementation that used it multiple times.